Aztec/Classes/Converters/NSAttributedStringToNodes.swift

```diff
@@ -61,31 +61,251 @@ class NSAttributedStringToNodes: Converter {
+
+            let branch = Branch(nodes: styleNodes, leaves: leafNodes)
+            branches.append(branch)
+        }
+
+        let paragraphNodes = createParagraphNodes(from: paragraph)
+        let processedBranches = process(branches: branches)
+
+        return reduce(nodes: paragraphNodes, leaves: processedBranches)
+    }
+}
+
+
+// MARK: - Merge: Helpers
+//
+private extension NSAttributedStringToNodes {
+
+    /// Defines a Tree Branch: Collection of Nodes, with a set of Leaves
+    ///
+    typealias Branch = (nodes: [ElementNode], leaves: [Node])
+
+
+    /// Defines a pair of Nodes that can be merged
+    ///
+    typealias MergeablePair = (left: ElementNode, right: ElementNode)
+
+
+    /// Sets Up a collection of Nodes and Leaves as a chain of Parent-Children, and returns the root node.and
+    /// If the collection of nodes is empty, will return the leaves parameters 'as is'.
+    ///
+    func reduce(nodes: [ElementNode], leaves: [Node]) -> [Node] {
+        return nodes.reduce(leaves) { (result, node) in
+            node.children = result
+            return [node]
+        }
+    }
+
+
+    /// Finds the Deepest node that can be merged "Right to Left", and returns the Left / Right matching touple, if any.
+    ///
+    func findMergeableNodes(left: [ElementNode], right: [ElementNode], blocklevelEnforced: Bool = true) -> [MergeablePair]? {
+        var currentIndex = 0
+        var matching = [MergeablePair]()
+
+        while currentIndex < left.count && currentIndex < right.count {
+            let left = left[currentIndex]
+            let right = right[currentIndex]
+
+            guard left.canMergeChildren(of: right, blocklevelEnforced: blocklevelEnforced) else {
+                break
+            }
+
+            let pair = MergeablePair(left: left, right: right)
+            matching.append(pair)
+            currentIndex += 1
+        }
+
+        return matching.isEmpty ? nil : matching
+    }
+
+
+    /// Debug Helper. Disregard!
+    ///
+    func print(node: Node, indentation: String = "") {
+        NSLog(indentation + "Node \(node.name) \(ObjectIdentifier(node))")
+
+        switch node {
+        case let text as TextNode:
+            NSLog(indentation + "Text.Content \(text.contents)")
+
+        case let element as ElementNode:
+            NSLog(indentation + "Element.Children:")
+
+            let deeperIndentation = indentation + String(.space) + String(.space)
+            for child in element.children {
+                print(node: child, indentation: deeperIndentation)
+            }
+
+        default:
+            break
+        }
+    }
+}
+
+
+// MARK: - Merge: Styles
+//
+private extension NSAttributedStringToNodes {
+
+    /// Given a collection of branches, this helper will iterate branch by branch and will:
+    ///
+    /// A. Reduce the Nodes: An actuall Parent/Child relationship will be set
+    /// B. Attempt to merge the current Branch with the Left-most-branch
```

Comment:
/s/left-most-branch/previous branch

"Leftmost" would always be the first branch.